### ISSUE TITLE
dynamic inventory plugin  request

### ISSUE THREAD

**Issue by macthev** [ansible/awx#11249]
##### ISSUE TYPE
 - Feature Idea

##### SUMMARY

A feature to support adding dynamic inventory.

awx 19.4
kube 1.20

I have looked at the following and can run on the cli inside the container but it seems difficult to run or include a dynamic inventory to what seems a common requirement from the GUI? Or am I missing something. The gui doesn't seem to have script options like found in earlier version.

https://github.com/ansible/awx/issues/3113#

**Comment by macthev** [comment]
I found I can run a plugin via the repo in awx  (proxmox in my case)  by including with the project.  
Ideally, the plugins would be a seperate repo from the playbooks and perhaps reused.

**Comment by AlanCoding** [comment]
Do you mean you are looking for `community.general.proxmox`? To get even more specific, you mean this:

https://github.com/ansible-collections/community.general/blob/2d6816e11e1672df5b2aa485e8af9eaa45d7c5be/plugins/inventory/proxmox.py

Could you give more info about how you solved this, and what could be improved? Going by what I read here, I'm thinking that you did:

 - add a file like `my.proxmox.yml` to your source control
 - created an SCM inventory from that project
 - But now this is missing a step, because you need the `community.general` collection for it to work, which you might have..
   - used `ansible-builder` to create a new execution environment with `community.general`
   - added the collection to `requirements.yml` for the project or something like that

I would like to see a feature where we allow people to put in a custom inventory config for an arbitrary inventory plugin from an arbitrary collection. The reason we haven't is because we still have the stumbling block of assuring that the collection exists. It would also be good to _validate_ the fully-qualified inventory plugin name.

So I would call this something like "inventory sourced from execution environment", as opposed to inventory sourced from project.

**Comment by macthev** [comment]
Yes , I wanted to be able to use a plug-in but found I needed to add to the project which means that there is a license problem because the plug-in maybe under one lic and the playbook could be under a different license. Ideally it would be simple as to have seperate plug-in configuration for the inventories.

**Comment by AlanCoding** [comment]
> Ideally it would be simple as to have seperate plug-in configuration for the inventories.

I understand your problem, but I don't follow this. Could you elaborate about what you see as the ideal solution to the problem?

**Comment by macthev** [comment]
Is there a way to have a dynamic plug-in without putting it in the project src?

**Comment by AlanCoding** [comment]
No, there is not, aside from the supported sources like gce, EC2, Azure Resource Manager, etc.

I can easily envision an enhancement to allow this generally (it's almost trivial on the API side), but we would need to clarify scope. Most inventory plugins for cloud sources can work by getting routed to from the "auto" inventory plugin built into Ansible core. These are the ones that put parameters in YAML and must have the `plugin:` key that gives the name of the plugin (`community.general.proxmox` in your case). If an inventory plugin does not adhere to this convention, then I wouldn't consider allowing use of it, but virtually all community "cloud" inventory plugins I've seen do follow the convention.

This was discussed in some prior issues IIRC, but we hesitated because we were not sure if we would get a plugin listing from execution environments (EEs). In your case, you need `community.general`, so it would make the most sense to start with an EE and have the server return a list of inventory plugins available inside of the EE.

**Comment by AlanCoding** [comment]
@ffirg @chrismeyersfsu right now there's not much to get excited about here, but in the long term, we might be able to make progress towards getting rid of the individual source choices (ec2, gce, etc.). Proper ownership of those are in domain of the execution environment and content maintainers.

**Comment by macthev** [comment]
Thanks for your comments. I do have a workaround so happy with that.

**Comment by dandrzejewski** [comment]
@macthev would you mind sharing your workaround? I'm interested in this as well.
